### PR TITLE
Add OneLocBuild template to arcade

### PR DESCRIPTION
**Pull request by jonfortescue**
Tested in [sdk](https://dev.azure.com/dnceng/internal/_build/results?buildId=1034620&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76&j=6aa31087-d1b8-5ee1-af29-f0fe63353381) and [winforms](https://dev.azure.com/dnceng/internal/_build/results?buildId=1032440&view=results), working in both (winforms pending [this](https://dev.azure.com/ceapex/CEINTL/_git/SoftwareLocalization/pullrequest/143036) PR to OneLocBuild to update their NuGet; working proof of concept for that [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=1032181&view=results)).

### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/core-eng/tree/master/Documentation/Validation

### CODE REVIEW

File: eng/common/generate-locproject.ps1 [hunk 1]
```diff
@@ -0,0 +1,89 @@
+Param(
+    [string] $LanguageSet = 'VS_Main_Languages',        # Language set to be used in the LocProject.json
```
Comment (MattGal) [inline eng/common/generate-locproject.ps1:2]: Is there a page or something you could link with known values?  This default value smells of magic constants.

File: eng/common/generate-locproject.ps1 [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    [string] $CreateFile = "true",                      # Creates a LocProject.json when "true," otherwise generates a LocProject.json and compares it to one that already exists in the repo 
+    [string] $CreateNeutralXlfs = "true"                # Creates neutral xlf files. Only set to false when running locally
+)
+
+Set-StrictMode -Version 2.0
+$ErrorActionPreference = "Stop"
+
+$exclusionsFilePath = "$env:BUILD_SOURCESDIRECTORY\Localize\LocExclusions.json"
```
Comment (MattGal) [inline eng/common/generate-locproject.ps1:10]: Confused: you mention running locally, then immediately use an ADO Agent variable?

File: eng/common/generate-locproject.ps1 [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+Push-Location "$env:BUILD_SOURCESDIRECTORY" # push location for Resolve-Path -Relative to work
+$jsonFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY" | Where-Object { $_.FullName -Match "en\\strings.json" }
+$xlfFiles = @()
+
+$allXlfFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY\*\*.xlf"
+$langXlfFiles = @()
+if ($allXlfFiles.Length -gt 0) {
+    $isMatch = $allXlfFiles[0].FullName -Match "\.([\w-]+)\.xlf" # matches '[langcode].xlf'
```
Comment (MattGal) [inline eng/common/generate-locproject.ps1:24]: These two lines are super confusing, perhaps there's a powershell-ism I'm missing?

- $isMatch is assigned but never used
- $Matches is used, but never assigned.

File: eng/common/generate-locproject.ps1 (hunk 2 above)
Comment (MattGal) [reply eng/common/generate-locproject.ps1:10]: if it's actually meant to use locally, consider reading these early on and providing outside-ADO defaults if they're not defined.

Comment (chcosta) [inline eng/common/generate-locproject.ps1:10]: This appears to be exclusively runnable in ci despite the comment for `CreateNeutralXlfs` implying otherwise.  Can we make this runnable outside of ci while still making this primarily targeted there?  ie,

```powershell

if (-Not (Test-Path -Path $env:BUILD_SOURCESDIRECTORY))
{
   Write-Host "This script is only intended to run in CI.  If you are running outside of ci, then you need to specify the 'BUILD_SOURCESDIRECTORY' environment variable."
  exit 1
{
```

File: eng/common/generate-locproject.ps1 [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+Set-StrictMode -Version 2.0
+$ErrorActionPreference = "Stop"
+
+$exclusionsFilePath = "$env:BUILD_SOURCESDIRECTORY\Localize\LocExclusions.json"
+$exclusions = @{ Exclusions = @() }
+if (Test-Path -Path $exclusionsFilePath)
+{
+    $exclusions = Get-Content "$env:BUILD_SOURCESDIRECTORY\Localize\LocExclusions.json" | ConvertFrom-Json
```
Comment (chcosta) [inline eng/common/generate-locproject.ps1:14]: why not use the `exclusionsFilePath`, which you set above, variable for this string?  Also, should this be a parameter (which defaults to this path) instead of being strictly stuck to this path?

File: eng/common/generate-locproject.ps1 [hunk 5]
```diff
@@ -0,0 +1,89 @@
+Param(
+    [string] $LanguageSet = 'VS_Main_Languages',        # Language set to be used in the LocProject.json
+    [string] $CreateFile = "true",                      # Creates a LocProject.json when "true," otherwise generates a LocProject.json and compares it to one that already exists in the repo
```
Comment (chcosta) [inline eng/common/generate-locproject.ps1:3]: bool? or switch?

File: eng/common/generate-locproject.ps1 [hunk 6]
```diff
@@ -0,0 +1,89 @@
+Param(
+    [string] $LanguageSet = 'VS_Main_Languages',        # Language set to be used in the LocProject.json
+    [string] $CreateFile = "true",                      # Creates a LocProject.json when "true," otherwise generates a LocProject.json and compares it to one that already exists in the repo 
+    [string] $CreateNeutralXlfs = "true"                # Creates neutral xlf files. Only set to false when running locally
```
Comment (chcosta) [inline eng/common/generate-locproject.ps1:4]: bool?

File: eng/common/generate-locproject.ps1 [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                        Remove-Item -Path $sourceFile
+                    }
+                }
+            )
+        }
+    )
+}
+
+$json = ConvertTo-Json $locJson -Depth 5
+Write-Host "(NETCORE_ENGINEERING_TELEMETRY=Build) LocProject.json generated:`n`n$json`n`n"
```
Comment (chcosta) [inline eng/common/generate-locproject.ps1:70]: Shouldn't be explicitly writing this string.  Should be importing pipelinelogging.ps1 and using the functions provided there.

File: eng/common/templates/job/onelocbuild.yml [hunk 8]
```diff
@@ -0,0 +1,71 @@
+parameters:
+  # Optional: dependencies of the job
+  dependsOn: ''
+
+  # Optional: A defined YAML pool - https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#pool
+  pool:
+    vmImage: vs2017-win2016
+
+  CeapexPat: $(dn-bot-ceapex-package-r)
```
Comment (chcosta) [inline eng/common/templates/job/onelocbuild.yml:9]: what's a ceapex?

File: eng/common/templates/job/onelocbuild.yml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+- job: OneLocBuild
+  
+  dependsOn: ${{ parameters.dependsOn }}
+
+  displayName: OneLocBuild
+
+  pool: ${{ parameters.pool }}
+
+  variables:
+    - group: OneLocBuildVariables
```
Comment (chcosta) [inline eng/common/templates/job/onelocbuild.yml:29]: Which secrets does this group provide?  what is defining the secrets that are provided to the parameters?

File: eng/common/templates/job/onelocbuild.yml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  steps:
+    - task: Powershell@2
```
Comment (chcosta) [inline eng/common/templates/job/onelocbuild.yml:32]: https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#pwsh is cleaner, but less controllable over what version of the task to use.  I'm not certain what our guidance is here but we should try to be consistent.

File: eng/common/templates/jobs/jobs.yml [hunk 11]
```diff
@@ -31,6 +31,14 @@ parameters:
   #           if 'true', the build won't run any of the internal only steps, even if it is running in non-public projects.
   runAsPublic: false
 
+  enableOneLocBuild: false
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml:34]: does this file need to know about any of the loc variables.  Today, this file is relatively generic.  At a minimum, can we do something slightly cleaner and use an object syntax like `graphFileGeneration` instead of just a flat list with gobs of parameters?

File: eng/common/generate-locproject.ps1 (hunk 2 above)
Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:10]: I does feel like perhaps these could just be parameters, and let the YAML pass them in.  I don't like when scripts read a bunch of values out of the environment, if avoidable. It makes it hard to follow the flow of data.

File: eng/common/templates/job/onelocbuild.yml [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        locProj: Localize/LocProject.json
+        outDir: $(Build.ArtifactStagingDirectory)
+        lclSource: ${{ parameters.LclSource }}
+        lclPackageId: ${{ parameters.LclPackageId }}
+        isCreatePrSelected: ${{ parameters.CreatePr }}
+        repoType: ${{ parameters.RepoType }}
+        gitHubPatVariable: "${{ parameters.GithubPat }}"
+        packageSourceAuth: patAuth
+        patVariable: ${{ parameters.CeapexPat }}
+      continueOnError: true
```
Comment (chcosta) [inline eng/common/templates/job/onelocbuild.yml:52]: Is this supposed to not cause failures?

File: eng/common/generate-locproject.ps1 [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+Push-Location "$env:BUILD_SOURCESDIRECTORY" # push location for Resolve-Path -Relative to work
+$jsonFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY" | Where-Object { $_.FullName -Match "en\\strings.json" }
+$xlfFiles = @()
+
+$allXlfFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY\*\*.xlf"
+$langXlfFiles = @()
+if ($allXlfFiles.Length -gt 0) {
+    $isMatch = $allXlfFiles[0].FullName -Match "\.([\w-]+)\.xlf" # matches '[langcode].xlf'
+    $firstLangCode = $Matches.1
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:25]: I've never seen the `.1` syntax... I would have expected `$match[1]`

File: eng/common/generate-locproject.ps1 [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    $exclusions = Get-Content "$env:BUILD_SOURCESDIRECTORY\Localize\LocExclusions.json" | ConvertFrom-Json
+}
+
+Push-Location "$env:BUILD_SOURCESDIRECTORY" # push location for Resolve-Path -Relative to work
+$jsonFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY" | Where-Object { $_.FullName -Match "en\\strings.json" }
+$xlfFiles = @()
+
+$allXlfFiles = Get-ChildItem -Recurse -Path "$env:BUILD_SOURCESDIRECTORY\*\*.xlf"
+$langXlfFiles = @()
+if ($allXlfFiles.Length -gt 0) {
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:23]: Powershell checks for empty listy things well. :-)
```suggestion
if ($allXlfFiles) {
```

File: eng/common/generate-locproject.ps1 (hunk 13 above)
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:25]: What happens if there is more than one "langcode" in the repo (maybe for test assets or soemthing).  It seems like the code should at least produce some sort of warning in that situation, since it's going to pick one arbitrarily at that point (and potentially not even deterministically, since we aren't sorting anything)

File: eng/common/generate-locproject.ps1 [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+$locFiles = $jsonFiles + $xlfFiles
+
+$locJson = @{
+    Projects = @(
+        @{
+            LanguageSet = $LanguageSet
+            LocItems = @(
+                $locFiles | ForEach-Object {
+                    $outputPath = "Localize\$(($_.DirectoryName | Resolve-Path -Relative).Substring(2) + "\")"
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:43]: Why "substring(2)"?  The "relative" is a little scary too, since its relative to wherever this script happens to be running from.  There should probably be a "Set-Location" in the script to put it somewhere reliable first. (Or Push-Location/Pop-Location)

File: eng/common/generate-locproject.ps1 [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+$locJson = @{
+    Projects = @(
+        @{
+            LanguageSet = $LanguageSet
+            LocItems = @(
+                $locFiles | ForEach-Object {
+                    $outputPath = "Localize\$(($_.DirectoryName | Resolve-Path -Relative).Substring(2) + "\")" 
+                    $continue = $true
+                    $exclusions.Exclusions | ForEach-Object {
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:45]: It's odd to use | ForEach-Object and not use the return.  I feel like
```ps
foreach ($exclusion in $exclusions.Exclusions) {
  # The code
}
```
makes it a little clearer that you are just looping over them, not transforming them in some way.  Right now it's going to return an array of $null, which is... odd. :-)

File: eng/common/generate-locproject.ps1 [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+                        if ($outputPath.Contains($_))
+                        {
+                            $continue = $false
+                        }
+                    }
+                    if ($continue)
+                    {
+                        $sourceFile = ($_.FullName | Resolve-Path -Relative).Substring(2)
+                        @{
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:54]: This should be returned.  It's REALLY confusing if it's not used... honestly, foreach-object is confusing enough without introducing the possibility of _multiple_ returns from it (or continuing after the code is returned)

File: eng/common/generate-locproject.ps1 (hunk 15 above)
Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:43]: Oh! There is a push, it's just a lot higher, never mind that.  Thought the "2" is still a little special.

File: eng/common/generate-locproject.ps1 [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    )
+}
+
+$json = ConvertTo-Json $locJson -Depth 5
+Write-Host "(NETCORE_ENGINEERING_TELEMETRY=Build) LocProject.json generated:`n`n$json`n`n"
+Pop-Location
+
+if ($CreateFile -eq "true") {
+    New-Item "$env:BUILD_SOURCESDIRECTORY\Localize\LocProject.json" -Force
```
Comment (ChadNedzlek) [inline eng/common/generate-locproject.ps1:74]: I think the "New-Item" is superfluous... "Set-Content" will create the file if it needs to

File: eng/common/generate-locproject.ps1 (hunk 5 above)
Comment (adiaaida) [reply eng/common/generate-locproject.ps1:3]: +1

File: eng/common/generate-locproject.ps1 (hunk 1 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:2]: Yessss I will link to the documentation.

File: eng/common/generate-locproject.ps1 (hunk 2 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:10]: That's a good idea. I will parameterize these.

File: eng/common/generate-locproject.ps1 (hunk 3 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:24]: Powershell-ism that's confusing but yeah

File: eng/common/generate-locproject.ps1 (hunk 2 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:10]: Making it runnable outside of CI

File: eng/common/generate-locproject.ps1 (hunk 4 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:14]: oh whoops, just missed that I guess. Will change and parameterize

File: eng/common/generate-locproject.ps1 (hunk 7 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:70]: oh didn't know that, thank you. Will do.

File: eng/common/templates/job/onelocbuild.yml (hunk 8 above)
Comment (jonfortescue) [reply eng/common/templates/job/onelocbuild.yml:9]: AzDO instance. Will link in comment.

File: eng/common/generate-locproject.ps1 (hunk 5 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:3]: switched these to switches

File: eng/common/generate-locproject.ps1 (hunk 6 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:4]: switch time

File: eng/common/templates/job/onelocbuild.yml (hunk 9 above)
Comment (jonfortescue) [reply eng/common/templates/job/onelocbuild.yml:29]: Added comment explaining.

File: eng/common/generate-locproject.ps1 (hunk 13 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:25]: Turns out that it doesn't matter -- it just needs *a* file with a langcode since all xlfs also contain English text equivalents, so we just take the first one and make it the "generic" xlf. This solution is actually ripped directly from their docs: https://ceapex.visualstudio.com/CEINTL/_wiki/wikis/CEINTL.wiki/1450/OneLocBuild-Non-Enu-source-file-support-(workaround)

Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:25]: But this implies that every set of "xlf" files in the entire repository have the same set of languages.  What if I have `a.en.xlf`, and `b.fr.xlf`.

This meands that `b.fr.xlf`, because we are only going to look for `*.en.xlf`.

Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:25]: Is that first sentence going to be something this code assumes, but doesn't validate?  That seems... scary... And validating it shouldn't be too hard... just search for ALL xlf files, and then group them by the first part.

Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:25]: Theoretically "dotnet script" is a thing. :-) Just sayin...

Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:25]: I'm fairly certain that it's a valid assumption to make because the LanguageSet variable applies to the entire repository -- all xlfs will be translated into the same set of languages.

I agree that the second solution is better but it's pretty difficult in powershell... that's exactly what I would have done had this been C# haha.

File: eng/common/generate-locproject.ps1 (hunk 15 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:43]: It removes the initial ".\" which matches the expected format for LocProject.json haha. actually, tho, it might work even with the .\ so let me try doing that.

File: eng/common/generate-locproject.ps1 (hunk 13 above)
Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1:25]: Ouch, quite the assumption for the tool to make, especially with repositories as large as ours. I hope that doesn't bite us. It... seems to hold for the repos I happen to have on my box? :-)

File: eng/common/generate-locproject.ps1 (hunk 15 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1:43]: Leaving the substring(2) out works fine, so i'm switching to that

File: eng/common/templates/jobs/jobs.yml (hunk 11 above)
Comment (jonfortescue) [reply eng/common/templates/jobs/jobs.yml:34]: removed this entirely

File: eng/common/generate-locproject.ps1 (hunk 15 above)
Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1]: Huzzah for weird path mechanics! :-)

**Comment by jonfortescue** [comment]
Delaying on merging this per an update I got from the templating team on Friday -- going to integrate some of their work before I merge.

File: eng/common/generate-locproject.ps1 (hunk 18 above)
Comment (jonfortescue) [reply eng/common/generate-locproject.ps1]: It turns out the New-Item is necessary bc Set-Content doesn't create directories (":

File: eng/common/templates/job/onelocbuild.yml (hunk 12 above)
Comment (jonfortescue) [reply eng/common/templates/job/onelocbuild.yml]: Changed to make it cause failures

**Comment by jonfortescue** [comment]
Should be ready to merge now that I got info from the templating team!

File: eng/common/generate-locproject.ps1 (hunk 18 above)
Comment (ChadNedzlek) [reply eng/common/generate-locproject.ps1]: Oooo, the "Localize" is the part we are really trying to create here. :-) I'd probably leave a comment to make sure someone doesn't helpfully remove it in the future. ;-)

Comment (jonfortescue) [reply eng/common/generate-locproject.ps1]: good idea!

**Review by ChadNedzlek — APPROVED** [review]

**Review by MattGal — APPROVED** [review]

**Comment by jonfortescue** [comment]
Successful runs in [winforms](https://dev.azure.com/dnceng/internal/_build/results?buildId=1042681&view=results) and [sdk](https://dev.azure.com/dnceng/internal/_build/results?buildId=1042692&view=results).

**Review by alexperovich — APPROVED** [review]
